### PR TITLE
[GraphBolt] Add DistributedItemSampler to support multi-gpu training

### PR DESCRIPTION
**Pull request by RamonZhou**
## Description
Create `DistributedItemSampler` as a replacement choice of `ItemSampler`. It will distribute inputs to each replica, supporting DDP training.
Added new params:
- `even_inputs`: It true, `DistributedItemSampler` will drop the uneven inputs so the number of batches is all the same for all replicas. Could be used in lightning. If False, user needs to use `Join` to avoid possible hanging.
- `num_replicas`: As the other torch APIs, user can choose to pass the number of replicas. If None, `DistributedItemSampler` will get this value from `dist`.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- `DistributedItemSampler`
- Unit test.

### CODE REVIEW

File: python/dgl/graphbolt/item_sampler.py [hunk 1]
```diff
@@ -326,8 +329,104 @@ def __iter__(self) -> Iterator:
...
+    """Distributed Item Sampler.
+
+    This is useful with torch's DDP training. It will create distributed item
+    subset of data which could be node IDs, node pairs with or without labels,
+    node pairs with negative sources/destinations, DGLGraphs and heterogeneous
+    counterparts. The items of the original item set will be sharded and form
+    an exclusive subset for each replica (process).
+
+    Note: This class `DistributedItemSampler` is not decorated with
+    `torchdata.datapipes.functional_datapipe` on purpose. This indicates it
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:348]: I can still see functional_datapipe decoration, conflict?

File: python/dgl/graphbolt/item_sampler.py [hunk 2]
```diff
@@ -326,8 +329,104 @@ def __iter__(self) -> Iterator:
...
+
+@functional_datapipe("distributed_sample_item")
+class DistributedItemSampler(ItemSampler):
+    """Distributed Item Sampler.
+
+    This is useful with torch's DDP training. It will create distributed item
+    subset of data which could be node IDs, node pairs with or without labels,
+    node pairs with negative sources/destinations, DGLGraphs and heterogeneous
+    counterparts. The items of the original item set will be sharded and form
+    an exclusive subset for each replica (process).
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:345]: Powered by Bard:

This sampler creates a distributed subset of items from the given data set, which can be used for training with PyTorch's Distributed Data Parallel (DDP). The items can be node IDs, node pairs with or without labels, node pairs with negative sources/destinations, DGLGraphs, or heterogeneous counterparts. The original item set is sharded such that each replica (process) receives an exclusive subset.

File: python/dgl/graphbolt/item_sampler.py [hunk 3]
```diff
@@ -326,8 +329,104 @@ def __iter__(self) -> Iterator:
...
+        self._even_inputs = even_inputs
+        # Apply a sharding filter to distribute the items.
+        self._item_set = self._item_set.sharding_filter()
+        # Get world size.
+        if num_replicas is None:
+            if not dist.is_available():
+                raise RuntimeError(
+                    "Requires distributed package to be available"
+                )
+            num_replicas = dist.get_world_size()
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:400]: What if input num_replicas is conflict with dist.get_world_size()?

File: python/dgl/graphbolt/item_sampler.py [hunk 4]
```diff
@@ -326,8 +329,108 @@ def __iter__(self) -> Iterator:
...
+    minibatcher : Optional[Callable]
+        A callable that takes in a list of items and returns a `MiniBatch`.
+    drop_last : bool
+        Option to drop the last batch if it's not full.
+    shuffle : bool
+        Option to shuffle before sample.
+    num_replicas: int
+        Number of processes participating in distributed training. By default,
+        `world_size` is retrieved from the current distributed group.
+    even_inputs : bool
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:367]: even_inputs is not very descriptive, what about drop_uneven_batches?
```suggestion
    drop_uneven_batches : bool
```

File: python/dgl/graphbolt/item_sampler.py [hunk 5]
```diff
@@ -326,8 +329,108 @@ def __iter__(self) -> Iterator:
...
+        Option to make sure the numbers of batches for each replica are the
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:368]: Option to ensure that all replicas receive the same number of batches. If some replicas have more batches than the others, the redundant batches of those replicas will be dropped. If the drop_last parameter is also set to True, the last batch will be dropped before the redundant batches are dropped.
Note: When using Distributed Data Parallel (DDP) training, it is recommended to use the Join Context Manager provided by PyTorch to solve the uneven inputs problem. However, this option can be used if the Join Context Manager is not helpful for any reason.

File: python/dgl/graphbolt/item_sampler.py [hunk 6]
```diff
@@ -326,8 +329,108 @@ def __iter__(self) -> Iterator:
...
+        same. If some of the processes have more batches than the others, the
+        redundant batches of those processes will be dropped. If drop_last and
+        even_inputs are both true, it will execute drop_last before even_inputs.
+        When using DDP training, using Join Context Manager provided by pytorch
+        is the recommended way to solve the uneven inputs problem. But if Join
+        is not helpful due to any reason, you could try this option.
+
+    Examples
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:376]: Remove if no example is provided.

File: python/dgl/graphbolt/item_sampler.py [hunk 7]
```diff
@@ -326,8 +329,108 @@ def __iter__(self) -> Iterator:
...
+        Data to be sampled.
+    batch_size : int
+        The size of each batch.
+    minibatcher : Optional[Callable]
+        A callable that takes in a list of items and returns a `MiniBatch`.
+    drop_last : bool
+        Option to drop the last batch if it's not full.
+    shuffle : bool
+        Option to shuffle before sample.
+    num_replicas: int
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:364]: Is replica a good terminology to describe the processes?
num_processes

if you prefer num_replicas, then update the python doc:
`Number of model replicas created in Distributed Data Parallel (DDP) training.`

Or just world_size as used in the python doc.

File: python/dgl/graphbolt/item_sampler.py [hunk 8]
```diff
@@ -326,8 +329,108 @@ def __iter__(self) -> Iterator:
...
+        Number of processes participating in distributed training. By default,
+        `world_size` is retrieved from the current distributed group.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:366]: The number of model replicas that will be created during Distributed Data Parallel (DDP) training. By default, the world size is retrieved from the current distributed group.

File: python/dgl/graphbolt/item_sampler.py [hunk 9]
```diff
@@ -326,8 +329,108 @@ def __iter__(self) -> Iterator:
...
+    even_inputs : bool
+        Option to make sure the numbers of batches for each replica are the
+        same. If some of the processes have more batches than the others, the
+        redundant batches of those processes will be dropped. If drop_last and
+        even_inputs are both true, it will execute drop_last before even_inputs.
+        When using DDP training, using Join Context Manager provided by pytorch
+        is the recommended way to solve the uneven inputs problem. But if Join
+        is not helpful due to any reason, you could try this option.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:374]: What may happen if I did not use Join Context Manager to even the input, and also did not set this option to True?

Comment (frozenbugs) [reply python/dgl/graphbolt/item_sampler.py:374]: What may happen if I did not set drop_last and but set this option to True, in this case the last batches still will not have complete size.

File: python/dgl/graphbolt/item_sampler.py [hunk 10]
```diff
@@ -326,8 +329,108 @@ def __iter__(self) -> Iterator:
...
+        # Get world size.
+        if num_replicas is None:
+            if not dist.is_available():
+                raise RuntimeError(
+                    "Requires distributed package to be available"
+                )
+            num_replicas = dist.get_world_size()
+        total_len = len(item_set)
+        # Calculate the number of batches for each replica.
+        self._num_batches = total_len // (num_replicas * batch_size) + (
```
Comment (frozenbugs) [inline python/dgl/graphbolt/item_sampler.py:403]: The _num_batches is used as the min_num_batches across each process, is this formula correct in calculate the min_num_batches?

File: tests/python/pytorch/graphbolt/test_item_sampler.py [hunk 11]
```diff
@@ -610,3 +614,110 @@ def test_ItemSetDict_node_pairs_negative_dsts(batch_size, shuffle, drop_last):
...
+    num_batches = len(list(item_sampler))
+
+    # Calculate expected numbers of items and batches.
+    expected_num_items = num_ids // nprocs + (num_ids % nprocs > proc_id)
+    if drop_last and expected_num_items % batch_size > 0:
+        expected_num_items -= expected_num_items % batch_size
+    expected_num_batches = expected_num_items // batch_size + (
+        (not drop_last) and (expected_num_items % batch_size > 0)
+    )
+    if even_inputs:
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_item_sampler.py:664]: Consider to test whether this option can actually even the input in different cases.

File: python/dgl/graphbolt/item_sampler.py (hunk 1 above)
Comment (RamonZhou) [reply python/dgl/graphbolt/item_sampler.py:348]: This is the same as ItemSampler (L90). I think it means it's not added **on purpose**.

Comment (frozenbugs) [reply python/dgl/graphbolt/item_sampler.py:348]: @Rhett-Ying can you clarify the conflict?

Comment (Rhett-Ying) [reply python/dgl/graphbolt/item_sampler.py:349]: I have removed the functional decorator in #6357. please remove it too

File: python/dgl/graphbolt/item_sampler.py [hunk 12]
```diff
@@ -326,8 +329,114 @@ def __iter__(self) -> Iterator:
...
+        self._drop_uneven_inputs = drop_uneven_inputs
+        # Apply a sharding filter to distribute the items.
+        self._item_set = self._item_set.sharding_filter()
+        # Get world size.
+        if num_replicas is None:
+            assert (
+                dist.is_available()
+            ), "Requires distributed package to be available"
+            num_replicas = dist.get_world_size()
+        total_len = len(item_set)
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/item_sampler.py:407]: `__len__` is not always available though it's true for most cases. how to handle this scenario?

File: python/dgl/graphbolt/item_sampler.py [hunk 13]
```diff
@@ -326,8 +329,114 @@ def __iter__(self) -> Iterator:
...
+        # Calculate the number of batches for each replica.
+        self._num_batches = total_len // (num_replicas * batch_size) + (
+            (not drop_last)
+            and (total_len % (num_replicas * batch_size) >= num_replicas)
         )
 
+    def __iter__(self) -> Iterator:
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/item_sampler.py:414]: below code is quite like to the implementation of `ItemSampler`. is it possible to remove such code duplication? In current way, I don't see much benefit of subclassing `ItemSampler`.

File: tests/python/pytorch/graphbolt/test_item_sampler.py [hunk 14]
```diff
@@ -610,3 +614,121 @@ def test_ItemSetDict_node_pairs_negative_dsts(batch_size, shuffle, drop_last):
...
+
+    # Create a DistributedItemSampler.
+    item_sampler = gb.DistributedItemSampler(
+        item_set,
+        batch_size=batch_size,
+        shuffle=shuffle,
+        drop_last=drop_last,
+        drop_uneven_inputs=drop_uneven_inputs,
+    )
+    data_loader = gb.SingleProcessDataLoader(item_sampler)
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/test_item_sampler.py:650]: what about `MultiProcessDL`? cannot be used?

File: python/dgl/graphbolt/item_sampler.py [hunk 15]
```diff
@@ -326,8 +329,114 @@ def __iter__(self) -> Iterator:
...
+        minibatcher: Optional[Callable] = minibatcher_default,
+        drop_last: Optional[bool] = False,
+        shuffle: Optional[bool] = False,
+        num_replicas: Optional[int] = None,
+        drop_uneven_inputs: Optional[bool] = False,
+    ) -> None:
+        super().__init__(item_set, batch_size, minibatcher, drop_last, shuffle)
+        self._drop_uneven_inputs = drop_uneven_inputs
+        # Apply a sharding filter to distribute the items.
+        self._item_set = self._item_set.sharding_filter()
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/item_sampler.py:400]: so the worflow in `DistribuedItemSampler` is `shard-> shuffle -> batch -> ...` ?

File: python/dgl/graphbolt/item_sampler.py (hunk 3 above)
Comment (RamonZhou) [reply python/dgl/graphbolt/item_sampler.py:406]: It will use the given `num_replicas`.

Comment (frozenbugs) [reply python/dgl/graphbolt/item_sampler.py:446]: Let's me make the question clear, in which case num_replicas does not equal to get_world_size.
Intuitively we don't need to pass in num_replicas since we can get_world_size. The comment should focus on address this confusion.

Comment (RamonZhou) [reply python/dgl/graphbolt/item_sampler.py:446]: It could be used when `dist` is not available during the instantiation of `DistributedItemSampler`. Most of the cases we don't have to use this param. But the comment is clear I think: we only retrieve it from `dist.get_world_size()` when `num_replicas` is None.

File: python/dgl/graphbolt/item_sampler.py (hunk 15 above)
Comment (RamonZhou) [reply python/dgl/graphbolt/item_sampler.py:440]: Yes

File: tests/python/pytorch/graphbolt/test_item_sampler.py (hunk 14 above)
Comment (RamonZhou) [reply tests/python/pytorch/graphbolt/test_item_sampler.py]: Could be used but not perfect.

File: python/dgl/graphbolt/item_sampler.py (hunk 9 above)
Comment (RamonZhou) [reply python/dgl/graphbolt/item_sampler.py]: The `drop_last` cares only about the completeness of the last batch, while `drop_uneven_inputs` cares only about the number of batches. I've added example cases of these situations.

File: python/dgl/graphbolt/item_sampler.py (hunk 10 above)
Comment (RamonZhou) [reply python/dgl/graphbolt/item_sampler.py]: It's only used when `drop_uneven_inputs` is True and it's used for limiting the number of batches in the datapipe. It would be fine if tests are all passed.

File: python/dgl/graphbolt/item_sampler.py [hunk 16]
```diff
@@ -306,27 +325,147 @@ def _collate(batch):
...
-        )
+        data_pipe = data_pipe.map(partial(self._minibatcher, names=self._names))
 
         return iter(data_pipe)
+
+
+class DistributedItemSampler(ItemSampler):
+    """Distributed Item Sampler.
+
+    This sampler creates a distributed subset of items from the given data set,
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/item_sampler.py:343]: I think we'd better clearly introduce the workflow: `shard -> shuffle -> ...` to help user have a clear understanding of this datapipe. And For this dist one, add note for the shuffle part which may lower the data diversity potentially.

File: python/dgl/graphbolt/item_sampler.py (hunk 12 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/item_sampler.py]: This is still a valid question, maybe we can not support multi-gpu training if we can't get the len of item_set.

@Rhett-Ying do you have other ideas?

**Review by frozenbugs — APPROVED** [review]
